### ISSUE TITLE
Operators and some `Merge` (e.g. `Add`, `Maximum`) layers do not propagate masks

### ISSUE THREAD

**Issue by jackd** [keras-team/keras#18416]
Masks are not propagated by add/mul operators, nor using some `Merge` layers (e.g. `Add`, `Maximum`). The below also demonstrates that this bug is also partially present in `tf.keras`  (operators fail, but `Add` / `Maximum` work as expected). Update: `Add` and `Maximum` no longer work in tf.keras.

Below is a simple example of learning the parameters of a straight line, `y = m * x + c`, where `(x, y)` are batched into shape `(batch_size, sequence_length, 1)` to mimic common usage e.g. in NLP. Masked examples have `x = y = 0`. We construct a model `y = dense(map_fn(x))`, where we use various `map_fn`s that should not change the capacity of the model (and in some cases should not even affect gradients). A summary of those that work "as expected" (i.e. learn roughly the correct weights) is shown below.

| map_fn | works with `keras` | works with `tf.keras` |
|----------|--------------------------|-----------------------|
| `lamda x: x` | &check;                    | &check;                  |
| `lambda x: x * 1` | &cross;            | &cross;                    |
| `lambda x: x + x - x` | &cross;            | &cross;               |
| `lambda x: x + x` | &cross;            | &cross;                    |
| `lambda x: keras.layers.Add()((x, x)` | &cross;            | &cross;                    |
| `lambda x: keras.layers.Maximum()((x, x)` | &cross;            | &cross;                    |
| `lambda x: AddSelf()(x)` | &check;            | &check;                    |

## example.py

```python
import sys
import typing as tp
import tensorflow as tf
USE_CORE = {"core": True, "tf": False}[sys.argv[1]]

if USE_CORE:
    import keras
    def get_value(x):
        return x.value
else:
    keras = tf.keras
    def get_value(x):
        return x

def get_dataset():
    """Create some masked dummy data, `y = gradient * x + offset`."""
    max_length = 16
    gradient = 5.
    offset = 2.
    batch_size = 64
    examples_per_epoch = 1000

    data_size = batch_size * examples_per_epoch
    x = tf.random.uniform((data_size, max_length))
    mask = (tf.ragged.range(
        tf.random.uniform((data_size,), 1, max_length, dtype=tf.int64)
    ) + 1).to_tensor(shape=(data_size, max_length)) > 0
    x = tf.where(mask, x, tf.zeros_like(x))
    y = tf.where(mask, x * gradient + offset, tf.zeros_like(x))

    x = tf.expand_dims(x, axis=-1)
    y = tf.expand_dims(y, axis=-1)

    dataset = tf.data.Dataset.from_tensor_slices((x, y))

    #### Adding sample_weight like below fixes the issue
    # sample_weight = tf.cast(mask, tf.float32)
    # dataset = tf.data.Dataset.from_tensor_slices((x, y, sample_weight))

    return dataset.batch(batch_size)


def get_model(input_shape, map_fn: tp.Callable):
    """Create model corresponding to `dense(map_fn(x))`."""
    inp = keras.Input(shape=input_shape[1:], batch_size=input_shape[0])
    x = keras.layers.Masking()(inp)
    x = map_fn(x)
    x = keras.layers.Dense(1)(x)
    return keras.Model(inp, x)
    


def build_and_eval(dataset: tf.data.Dataset, map_fn: tp.Callable, desc: str):
    """Build and evaluate a model dense(map_fn(x)) on dataset."""
    print("----------------")
    print(desc)
    print("----------------")
    model = get_model(dataset.element_spec[0].shape, map_fn)
    if hasattr(model.output, "_keras_mask"):
        print("Output has '_keras_mask' attr")
    else:
        print("Output does NOT have a 'keras_mask' attr")
    model.compile(loss="mse", optimizer="sgd")
    model.fit(dataset, epochs=2)
    print([(w.name, get_value(w).numpy().item()) for w in model.weights])


class AddSelf(keras.layers.Layer):
    def __init__(self):
        super().__init__()
        self.supports_masking = True
    
    def call(self, x):
        return x + x


def main():
    tf.random.set_seed(42)
    dataset = get_dataset()
    
    build_and_eval(dataset, lambda x: x, "x")
    build_and_eval(dataset, lambda x: x * 1, "x * 1")
    build_and_eval(dataset, lambda x: x + x - x, "x + x - x")
    build_and_eval(dataset, lambda x: x + x, "x + x")
    build_and_eval(
        dataset,
        lambda x: keras.layers.Add()((x, x)),
        "keras.layers.Add()((x, x))",
    )
    build_and_eval(
        dataset,
        lambda x: keras.layers.Maximum()((x, x)),
        "keras.layers.Maximum()((x, x))",
    )
    build_and_eval(dataset, lambda x: AddSelf()(x), "AddSelf()(x)")


if __name__ == '__main__':
    main()
```

## Using `keras`

```bash
python example.py core
```

```txt
----------------
x
----------------
Output has '_keras_mask' attr
Epoch 1/2
1000/1000 ━━━━━━━━━━━━━━━━━━━━ 2s 837us/step - loss: 0.6007
Epoch 2/2
1000/1000 ━━━━━━━━━━━━━━━━━━━━ 1s 834us/step - loss: 0.0459
[('kernel', 3.0166471004486084), ('bias', 2.8878185749053955)]
----------------
x * 1
----------------
Output does NOT have a 'keras_mask' attr
Epoch 1/2
1000/1000 ━━━━━━━━━━━━━━━━━━━━ 1s 893us/step - loss: 2.7253
Epoch 2/2
1000/1000 ━━━━━━━━━━━━━━━━━━━━ 1s 858us/step - loss: 0.4369
[('kernel', 7.281540870666504), ('bias', 0.4343944489955902)]
----------------
x + x - x
----------------
Output does NOT have a 'keras_mask' attr
Epoch 1/2
1000/1000 ━━━━━━━━━━━━━━━━━━━━ 1s 798us/step - loss: 2.4995
Epoch 2/2
1000/1000 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - loss: 0.4331
[('kernel', 7.2877912521362305), ('bias', 0.4326573312282562)]
----------------
x + x
----------------
Output does NOT have a 'keras_mask' attr
Epoch 1/2
1000/1000 ━━━━━━━━━━━━━━━━━━━━ 2s 1ms/step - loss: 1.9006
Epoch 2/2
1000/1000 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - loss: 0.3997
[('kernel', 3.7006561756134033), ('bias', 0.4010191261768341)]
----------------
keras.layers.Add()((x, x))
----------------
Output does NOT have a 'keras_mask' attr
Epoch 1/2
1000/1000 ━━━━━━━━━━━━━━━━━━━━ 2s 1ms/step - loss: 1.8395
Epoch 2/2
1000/1000 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - loss: 0.3997
[('kernel', 3.7006564140319824), ('bias', 0.40101900696754456)]
----------------
keras.layers.Maximum()((x, x))
----------------
Output does NOT have a 'keras_mask' attr
Epoch 1/2
1000/1000 ━━━━━━━━━━━━━━━━━━━━ 2s 1ms/step - loss: 2.6779
Epoch 2/2
1000/1000 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - loss: 0.4361
[('kernel', 7.282827377319336), ('bias', 0.4340367913246155)]
----------------
AddSelf()(x)
----------------
Output has '_keras_mask' attr
Epoch 1/2
1000/1000 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - loss: 0.4728
Epoch 2/2
1000/1000 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - loss: 0.0022
[('kernel', 2.5691280364990234), ('bias', 1.8994231224060059)]
```

## Using `tf.keras`

```bash
python example.py tf
```

```txt
----------------
x
----------------
Output has '_keras_mask' attr
Epoch 1/2
1000/1000 [==============================] - 4s 3ms/step - loss: 0.4196
Epoch 2/2
1000/1000 [==============================] - 3s 3ms/step - loss: 0.0084
[('dense/kernel:0', 4.857532501220703), ('dense/bias:0', 2.0762569904327393)]
----------------
x * 1
----------------
Output does NOT have a 'keras_mask' attr
Epoch 1/2
1000/1000 [==============================] - 5s 5ms/step - loss: 1.6487
Epoch 2/2
1000/1000 [==============================] - 4s 4ms/step - loss: 0.4224
[('dense_1/kernel:0', 7.2698187828063965), ('dense_1/bias:0', 0.4376528561115265)]
----------------
x + x - x
----------------
Output does NOT have a 'keras_mask' attr
Epoch 1/2
1000/1000 [==============================] - 6s 6ms/step - loss: 1.3704
Epoch 2/2
1000/1000 [==============================] - 7s 7ms/step - loss: 0.4173
[('dense_2/kernel:0', 7.285768032073975), ('dense_2/bias:0', 0.43321940302848816)]
----------------
x + x
----------------
Output does NOT have a 'keras_mask' attr
Epoch 1/2
1000/1000 [==============================] - 7s 7ms/step - loss: 0.9798
Epoch 2/2
1000/1000 [==============================] - 4s 4ms/step - loss: 0.3998
[('dense_3/kernel:0', 3.7006540298461914), ('dense_3/bias:0', 0.4010206162929535)]
----------------
keras.layers.Add()((x, x))
----------------
Output does NOT have '_keras_mask' attr
Epoch 1/2
1000/1000 [==============================] - 5s 5ms/step - loss: 0.4362
Epoch 2/2
1000/1000 [==============================] - 6s 6ms/step - loss: 3.6697e-05
[('dense_4/kernel:0', 2.498835563659668), ('dense_4/bias:0', 2.0013747215270996)]
----------------
keras.layers.Maximum()((x, x))
----------------
Output doe NOT have a '_keras_mask' attr
Epoch 1/2
1000/1000 [==============================] - 6s 6ms/step - loss: 0.4207
Epoch 2/2
1000/1000 [==============================] - 7s 7ms/step - loss: 0.0084
[('dense_5/kernel:0', 4.857173919677734), ('dense_5/bias:0', 2.076448917388916)]
----------------
AddSelf()(x)
----------------
Output has '_keras_mask' attr
Epoch 1/2
1000/1000 [==============================] - 6s 6ms/step - loss: 0.1189
Epoch 2/2
1000/1000 [==============================] - 6s 6ms/step - loss: 5.4465e-05
[('dense_6/kernel:0', 2.501418113708496), ('dense_6/bias:0', 1.9983253479003906)]
```

**Comment by sachinprasadhs** [comment]
@jackd , Is the issue still relevant? if so, you may need to rephrase the question about comparison since tf.keras also uses Keras 3 in backend.

**Comment by jackd** [comment]
@sachinprasadhs updated OP. Bugs now uniformly present across keras and tf.keras - some working cases no longer work, i.e. behaviour is now worse than when posted.

**Comment by hertschuh** [comment]
@jackd ,

Thanks for the report. There were indeed a few bugs that caused two issues:
- no mask was returned on the outputs
- the mask was not taken into account when performing the merge operation (for instance `Minimum` would take the minimum with zero outside of the mask).

I have a PR open to fix those. I will consider that the merge layers are the scope for this issue.

The other discussion points are much broader.

> I think everything would be a lot simpler if we had a MaskedTensor composite class, but that's a discussion for somewhere else...

The concept of `MaskedTensor`s doesn't exist in all backends, so this is not really an option at this time.

> it is expected that using low-level operators like `-` and `+` will not preserve masks

One of the unintuitive things about Keras supporting multiple backends is that `x + y` can behave differently than `ops.add(x, y)` in subtle ways. That is because `+` is handled natively by the backend (Tensorflow, JAX, Torch) without Keras having any visibility into it. Without native backend support for masks, `x + y` won't propagate masks.

> If these layers don't propagate masks, at the very least I would expect a warning about not propagating as with other layers that don't propagate.

This warning exists: https://github.com/keras-team/keras/blob/master/keras/src/layers/layer.py#L879-L885  It didn't fire in this case because the merge layers claim to support masking and do implement `compute_mask`. Simply, the implementation was wrong. Note that not having a mask on the output does not mean that the layer unintentionally lost the mask. For instance, it is expected that `Add()([x, y])` will have no mask if x has a mask and y has no mask.

**Comment by jackd** [comment]
@hertschuh 

> The concept of MaskedTensors doesn't exist in all backends

So keras only ever plans to implement composite tensors that are implemented in all backends? Getting a bit off topic for this issue, but that seems like an unnecessary restriction to me...

> That is because + is handled natively by the backend (Tensorflow, JAX, Torch)

Isn't [this](https://github.com/keras-team/keras/blob/master/keras/src/backend/common/keras_tensor.py#L130) how add is handled? Looks like any PR that fixes `Add` will also fix `__add__`.

> For instance, it is expected that Add()([x, y]) will have no mask if x has a mask and y has no mask.

That's not at all what I would expect. What is a masked tensor supposed to represent? My understanding is that it's essentially a different representation of tensorflow's `RaggedTensor`. Adding a `RaggedTensor` with shape `[N, ?]` to a regular tensor of shape `[N, M]` results in an error AFAIK. Are there any situations where it makes sense to add a masked and unmasked tensor?
